test/sqllogictest/chbench.slt

```diff
@@ -856,6 +855,8 @@ ORDER BY ordercount DESC
 
 %2 =
 | Get materialize.public.nation (u31)
+| Filter (#1 = "GERMANY")
+| Project (#0)
```

Comment:
It looks like this is a new arrangement, whereas before we hoisted the filter in order to re-use an existing arrangement. Is that on purpose, or a potential glitch?

Reply:
This is a glitch.

I investigated, and it turns out it is because that projection pushdown fights with Relation CSE.

----

An explanation with a simplified version of the query:

On main, our plan starts out like this:
```
%0
| Get stock

%1
| Get supplier

%2
| Get nation
| Filter (#1 = "GERMANY")

%3
| Join %0 %1 %2 <equivalences>
| Reduce group=(#0)
| | agg sum(#14)

%4
| Get stock

%5
| Get supplier 

%6
| Get nation
| Filter (#1 = "GERMANY")

%7
| Join %4 %5 %6 <equivalences>
| Reduce group=()
| | agg sum(#14)
```
and then RelationCSE finds it is the same join with two different reduces applied, so it consolidates the query like this:
```
%0
| Get stock

%1
| Get supplier

%2
| Get nation
| Filter (#1 = "GERMANY")

%3 = l0
| Join %0 %1 %2 <equivalences>

%4
| Get (l0)
| Reduce group=(#0)
| | agg sum(#14)

%5
| Get (l0)
| Reduce group=()
| | agg sum(#14)
```

However, with projection pushdown, one join has become one column narrower than the other.
```
%0
| Get stock
| Project(#0)

%1
| Get supplier
| Project(#0)

%2
| Get nation
| Filter (#1 = "GERMANY")

%3
| Join %0 %1 %2 <equivalences>
| Reduce group=(#0)
| | agg sum(#14)

%4
| Get stock
| Project() <-- Removes one extra column because the reduce has no group by keys.

%5
| Get supplier 
| Project(#0)

%6
| Get nation
| Filter (#1 = "GERMANY")

%7
| Join %4 %5 %6 <equivalences>
| Reduce group=()
| | agg sum(#14)
```

Thus, RelationCSE can only consolidate the same join input:
```
%0
| Get stock
| Project(#0)

%1
| Get supplier
| Project(#0)

%2 = l0
| Get nation
| Filter (#1 = "GERMANY")

%3
| Join %0 %1 %2 <equivalences>
| Reduce group=(#0)
| | agg sum(#14)

%4
| Get stock
| Project() 

%5
| Get supplier 
| Project(#0) <-- Why "supplier" does not get consolidated into a let is a mystery to me

%6
| Join %4 %5 %0 <equivalences>
| Reduce group=()
| | agg sum(#14)
```

Because the let statement, the `Filter (#1 = "GERMANY")` cannot be lifted by `JoinImplementation`. But `JoinImplementation` does lift the other projects, so `RelationCSE` recognizes that the two joins are the same after `JoinImplementation` has done its job.

Reply:
Well, that makes sense! I think this is an example of the sort of thing @asenac was asking about: a case where specialization can result in distinct joins that are hard to put back together (it's not a predicate based specialization, but still).

My hunch on this is that applying projection pushdown later is better; I'm not sure where you have it at the moment, but it has the "feel" of a physical optimization. That is, the actual deletion of the columns is a transformation that is more specific than the "demand" sort of analysis that says "keep the same shape, but feel welcome to ignore some values". Not that we need to keep `Demand` around, but that strikes me as the difference between the two.

I think this is the sort of thing we might expect to see a bit (joins, followed by different reductions) so it's probably worth sorting out what to do with it. Happy to chat more about it, but I'm curious if deferring projection pushdown to the dataflow stage (or beyond, to a pre-LIR thing like where JoinPlanning should be) makes sense to you?

Reply:
Report:
I moved projection pushdown to be the first of the physical optimizations. This fixed the problem above, but it made a certain form of stacked joins worse.
With delayed projection pushdown (which is what is shown below), the two joins get planned as Differential joins.
```
%0 = Let l0 =
| Get ? (u19)
| Project (#4, #7, #8)
| Filter !(isnull(#0))

%1 =
| Get ? (u24)
| Project (#0, #4)
| Filter "^.*b$" ~(padchar(#1))
| Project (#0)

%2 =
| Get %0 (l0)
| Project (#0, #1)

%3 =
| Join %1 %2 (= #0 #1)
| | implementation = Unimplemented
| Project (#0, #2)
| Reduce group=(#0)
| | agg sum(#1)
| | agg count(#1)
| Map (i64tof64(#1) / i64tof64(if (#2 = 0) then {null} else {#2}))
| Project (#0, #3)

%4 = Let l1 =
| Join %0 %3 (= #0 #3)
| | implementation = Unimplemented
| Project (#1, #2, #4)
| Filter (i32tof64(#0) < #2)
| Project (#1)
| Reduce group=()
| | agg sum(#0)
```

With the projection pushdown happening earlier, the stacked joins get planned as DeltaQueries because the joins are not considered to be using the same input:
```
%0 =
| Get ? (u19)
| Project (#4, #7)
| Filter !(isnull(#0))

%1 =
| Get ? (u24)
| Project (#0, #4)
| Filter "^.*b$" ~(padchar(#1))
| Project (#0)

%2 =
| Join %1 %0 (= #0 #1)
| | implementation = Unimplemented
| Project (#0, #2)
| Reduce group=(#0)
| | agg sum(#1)
| | agg count(#1)
| Map (i64tof64(#1) / i64tof64(if (#2 = 0) then {null} else {#2}))
| Project (#0, #3)

%3 =
| Get ? (u19)
| Project (#4, #7, #8)
| Filter !(isnull(#0))

%4 = Let l1 =
| Join %3 %2 (= #0 #3)
| | implementation = Unimplemented
| Project (#1, #2, #4)
| Filter (i32tof64(#0) < #2)
| Project (#1)
| Reduce group=()
| | agg sum(#0)
```

I have decided to add a field `inline_mfp` to the `InlineLet` transform so that calls to `InlineLet` that happen before `JoinImplementation` will inline expressions consisting of a MFP around a Get.